[MPS] Add support for Custom Kernels

razarmehr opened the pull request:
- This change introduces these APIs to enable developing custom kernels on the MPS Stream:
`torch::mps::get_command_buffer()`
`torch::mps::get_dispatch_queue()`
`torch::mps::commit()`
- Add ObjC test case

razarmehr committed:
[MPS] Add support for Custom Kernels
- This change introduces these APIs to enable developing custom kernels
on the MPS Stream:
`torch::mps::get_command_buffer()`
`torch::mps::get_dispatch_queue()`
`torch::mps::commit()`
- Add ObjC test case

razarmehr commented:
The CI trunk failures seem unrelated to the PR:
`bash: ./.ci/pytorch/python_doc_push_script.sh: No such file or directory`

albanD commented on torch/csrc/mps/Module.cpp:
```diff
@@ -74,7 +74,7 @@ static PyObject* MPSModule_isMacOS13orNewer(PyObject* _unused, PyObject* args) {
 static PyObject* MPSModule_synchronize(PyObject* _unused, PyObject* noargs) {
   HANDLE_TH_ERRORS
-  at::detail::getMPSHooks().deviceSynchronize();
+  at::detail::getMPSHooks().synchronizeStream();
```
huh? 
That doesn't seem to match the current doc for this function? https://pytorch.org/docs/stable/generated/torch.mps.synchronize.html?highlight=synchronize#torch.mps.synchronize

razarmehr replied:
We'll have `synchronizeEvent()` API coming in a separate PR, so I wanted to make the distinction with `synchronizeStream()`. We want Python API to remain as `torch.mps.synchronize()` to have the CUDA equivalent, and don't think it matters what our internal MPSHooks implementation calls it. 

I'll rename MPSModule_synchronize() to MPSModule_synchronizeStream() , though.

albanD replied:
The internal name doesn't matter much but it makes things easier to follow if they have the same name?
To me `deviceSynchronize` sounds like you synchronize the whole device (all streams there) while `synchronizeStream` is synchronizing a single stream (the current one here I guess).

razarmehr replied:
Renamed it back to `deviceSynchronize`.

albanD commented on caffe2/CMakeLists.txt:
```diff
@@ -1683,6 +1683,11 @@ if(BUILD_TEST)
       get_filename_component(test_name ${test_src} NAME_WE)
       add_executable(${test_name} "${test_src}")
       target_link_libraries(${test_name} torch_library gtest_main)
+      if(APPLE)
```
Huh, when you USE_MPS=1 but APPLE=0 ?
Or is this to single out mobile or something?

razarmehr replied:
This is to check if MacOS SDL libs such as Metal or Foundation are present. I'm not sure if it's ever possible to have  USE_MPS=1 on a non-Apple system, so I just wanted to be safe. 
If you think if whenever USE_MPS=1 then it's "always" an Apple OS, then I'll remove that condition and merge the  `target_link_libraries()`

albanD replied:
That feels like a fair assumption in my head yes.
We can add a small check above to make sure it is true once and for all as well if we're worried.

razarmehr replied:
Merged the two `target_link_libraries` and updated the PR.

albanD commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -17,5 +26,15 @@ void TORCH_API manual_seed(uint64_t seed);
 /// Waits for all streams on a MPS device to complete.
 void TORCH_API synchronize();
 
+/// Submits the command buffer to run on the GPU
```
I feel like these functions would need quite a bit more of documentation?

Also I'm very commit() doesn't take a command buffer or dispatch queue

razarmehr replied:
>I feel like these functions would need quite a bit more of documentation?

I can add links to Metal documentation.

>Also I'm very commit() doesn't take a command buffer or dispatch queue
At this time, there's only one active/current command buffer we have in MPSStream.
Also the dispatch_queue_t in this context is just a Mutex, and has nothing to do with GPU streams.

kulinseth replied:
> Also I'm very commit() doesn't take a command buffer or dispatch queue

Alban, commit() method works on the Current command buffer object which you can get from `get_command_buffer`. This is the commandBuffer which MPS backend is currently using and encoding work into. This is already using a commitAndContinue feature which allows for concurrency with CPU. Proving handle to the same CommandBuffer to the external users, provide them the same treatment we give as the rest of the op implementation. 
With this context in-mind we decided not to have the commandBuffer passed to commit was to avoid the issue of a user created Command buffer being passed to the commit() method. 
I want to point out, a user can get the device and the Command Queue from the currentCommandBuffer. This will allow them to create new Command buffers if they wanted to. We would like to avoid this mis-match of the CommandBuffers on our APIs.

razarmehr replied:
Updated the PR.

albanD reviewed (commented):
btw if the answer to a lot of my questions is "there is a single stream so this is not relevant", it should be made very explicit.

razarmehr committed:
Rename synchronizeStream() to deviceSynchronize()
- Also reverted a change in commitAndWait() that was made in MPSProfiler PR
and was causing a hang with Custom Kernels.

kulinseth reviewed (approved):
Thanks @albanD for the review .
And @razarmehr for the PR . The change looks good

kulinseth reviewed (approved):
Thanks @albanD for the review .
And @razarmehr for the PR . The change looks good

malfet commented on aten/src/ATen/detail/MPSHooksInterface.h:
```diff
@@ -17,47 +17,52 @@ class Context;
-
   virtual void setMemoryFraction(double /*ratio*/) const {
-    AT_ERROR("Cannot execute setMemoryFraction() without MPS backend.");
+    FAIL_MPSHOOKS_FUNC(__func__);
   }
```
```suggestion
    FAIL_MPSHOOKS_FUNC(__func__);
  }
  
  #undef FAIL_MPSHOOKS_FUNC
```

razarmehr replied:
Good catch, thanks.

malfet commented on torch/_C/__init__.pyi.in:
```diff
@@ -1440,7 +1440,7 @@ class _TensorBase(metaclass=_TensorMeta):
 
 # Defined in torch/csrc/mps/Module.cpp
-def _mps_synchronize() -> None: ...
+def _mps_deviceSynchronize() -> None: ...
```
Can you please elaborate why this change is needed? Also, it makes it different from similar `_cuda_syncrhonize`, isn't it?

razarmehr replied:
As I said `[here](https://github.com/pytorch/pytorch/pull/100661#discussion_r1186261524)`, the API `synchronizeEvent()` is coming in a separate PR.  So `_mps_synchronize()` alone doesn't imply what it's synchronizing (event, single stream, device, etc.). So i renamed it to clarify what it's synchronizing.

malfet reviewed (approved):
Looks good, but please undef the define in header. (And again, would be nice to submit as separate PR as it has nothing to do with adding support for customer Kernels, is it?
And not sure why rename  `_mps_synchronize` to ``_mps_DeviceSynchronize`

albanD commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -15,7 +24,19 @@ bool TORCH_API is_available();
 
 /// Waits for all streams on a MPS device to complete.
+/// See this link for more info:
+/// https://developer.apple.com/documentation/metal/mtlcommandbuffer/1443039-waituntilcompleted
```
Thanks for adding this link but this still feels quite confusing to me.
For example the names "stream" and "command buffer"/"dispatch queue" are used without defining the difference between them.


At a high level I would expect that:
- You have links on what these data structures are in the mps docs
- Have explanation on how these structures are linked to the concepts that exist on at::device
- For the function API, you either need to say: (defined based off MPS function) "this function is equivalent to waitUntilCompleted which is described in the mps doc here" or (defined based off pytorch concepts) "this function is synchronizing the GPU by waiting for all ops in all streams to be done". I think the second one is best as it improves consistency within PT and makes it simpler to understand what this API is doing for other PT maintainers.
- If you need a whole Note on the website or in a specific file that maps MPS names to PT names, we should do that and link it from the code. Within code, we use `Note [Some Name]` to write longer notes that are linked from multiple places for example https://github.com/search?q=repo%3Apytorch%2Fpytorch+%2FNote+%5C%5BMove+construction+for+RAII+guards+is+tricky%5C%5D%2F&type=code

razarmehr replied:
Thanks @albanD for the suggestions. Since these comments for C++ APIs and the ones in `mps.rst` for python bindings need a good refining/scrubbing, if you don't mind I'll update both of them in a separate PR soon and will add you to the review (a PR just for API docs). Sounds good to you?

albanD replied:
>  the ones in mps.rst for python bindings

This is for the python side yes. I think a Note in the C++ code would be very valuable though for these lower level APIs!
Or if they all make it to public APIs, then having them nicely documented at the python level (with c++ pointing to the python doc) sounds good.

I'm fine with doing this in a follow up. But I don't feel like I can properly review PRs like this one without such a doc. So I would expect this before I can help with more reviews.

razarmehr replied:
Sure, I'll talk to Kulin and will follow up on that with you.

razarmehr replied:
I updated the comments for `torch::mps::synchronize()`.

albanD reviewed (commented):
The change in itself sounds ok but I think we need to be a lot stricter on the documentation if we expect non-MPS maintainers to be able to use this / be able to review PRs touching MPS code.

razarmehr commented on torch/csrc/mps/Module.cpp:
```diff
@@ -120,7 +122,10 @@ static PyObject* MPSModule_driverAllocatedMemory(
 // cppcoreguidelines-avoid-c-arrays)
 static struct PyMethodDef _MPSModule_methods[] = {
-    {"_mps_synchronize", MPSModule_synchronize, METH_NOARGS, nullptr},
+    {"_mps_deviceSynchronize",
```
Two reasons:
1- I renamed it based on your comment @albanD : https://github.com/pytorch/pytorch/pull/100661#discussion_r1186463260
I thought you said, deviceSynchronize makes more sense in you comment.

2- As I said `[here](https://github.com/pytorch/pytorch/pull/100661#discussion_r1186261524)`, the API `synchronizeEvent()` is coming in a separate PR.  So `_mps_synchronize()` alone doesn't imply what it's synchronizing (event, single stream, device, etc.). So i renamed it to clarify what it' synchronizing.

razarmehr committed:
Update aten/src/ATen/detail/MPSHooksInterface.h

Co-authored-by: Nikita Shulga <nikita.shulga@gmail.com>

razarmehr committed:
Update aten/src/ATen/detail/MPSHooksInterface.h

Co-authored-by: albanD <desmaison.alban@gmail.com>

razarmehr commented:
@pytorchbot merge

razarmehr commented:
@pytorchbot merge -f

razarmehr commented:
@pytorchbot merge -f "all checks except the pre-existing lint-linux failure are green"

huydhn commented:
@pytorchbot revert -m 'Lint does fail on this change though, so please just rebase and reland the change' -c ignoredsignal

huydhn commented:
Well, the lint failure is just a simple one line fix, so I will do a forward fix instead

razarmehr commented:
> Well, the lint failure is just a simple one line fix, so I will do a forward fix instead

Thanks @huydhn and @ZainRizvi for taking care of this.  
I applied this [Commit Suggestion](https://github.com/pytorch/pytorch/pull/100661#discussion_r1186711550) directly. When I did that it wasn't visible on Github that there are trailing spaces.

ZainRizvi commented:
Understandable, especially since Dr. CI miscategorized the failure.

Fyi, if you run `lintrunner -a` on your local machine it'll not only catch the lint errors but also auto-fix many of them for you too

jeanschmidt commented:
This change is breaking some builds on Mac:

```
   -> 4:15: error: typedef redefinition with different types ('void *' vs 'struct dispatch_queue_s *')
   -> typedef void* dispatch_queue_t;
   ->               ^
   -> arvr/third-party/toolchains/xcode/14.1.0/MacOSX.sdk/usr/include/dispatch/queue.h:70:1: note: previous definition is here
   -> DISPATCH_DECL(dispatch_queue);
   -> ^
   -> arvr/third-party/toolchains/xcode/14.1.0/MacOSX.sdk/usr/include/dispatch/object.h:89:58: note: expanded from macro 'DISPATCH_DECL'
   ->                 typedef struct name##_s : public dispatch_object_s {} *name##_t
   ```
   
   due to the redefinition of `dispatch_queue_t` I believe it should be reverted, not forward fixed (note @huydhn #100907 )

razarmehr commented:
> This change is breaking some builds on Mac:
> 
> ```
>    -> 4:15: error: typedef redefinition with different types ('void *' vs 'struct dispatch_queue_s *')
>    -> typedef void* dispatch_queue_t;
>    ->               ^
>    -> arvr/third-party/toolchains/xcode/14.1.0/MacOSX.sdk/usr/include/dispatch/queue.h:70:1: note: previous definition is here
>    -> DISPATCH_DECL(dispatch_queue);
>    -> ^
>    -> arvr/third-party/toolchains/xcode/14.1.0/MacOSX.sdk/usr/include/dispatch/object.h:89:58: note: expanded from macro 'DISPATCH_DECL'
>    ->                 typedef struct name##_s : public dispatch_object_s {} *name##_t
> ```
> 
> due to the redefinition of `dispatch_queue_t` I believe it should be reverted, not forward fixed (note @huydhn #100907 )

That's weird. I added ciflow/trunk too. Why CIs didn't catch this.

jeanschmidt commented:
Could it be because some jobs got canceled? I see https://github.com/pytorch/pytorch/actions/runs/4919627966/jobs/8787795889 but not sure if this or others should catch this

jeanschmidt commented:
So, seems that it might be possible to reproduce with `ciflow/periodic` (thanks @huydhn). So I will revert it in this case.

kulinseth commented:
> Could it be because some jobs got canceled? I see https://github.com/pytorch/pytorch/actions/runs/4919627966/jobs/8787795889 but not sure if this or others should catch this

@huydhn , this issue should be resolved.?

huydhn commented:
> > Could it be because some jobs got canceled? I see https://github.com/pytorch/pytorch/actions/runs/4919627966/jobs/8787795889 but not sure if this or others should catch this
> 
> @huydhn , this issue should be resolved.?

Yes, please do a rebase (pass https://hud.pytorch.org/pytorch/pytorch/commit/4447dfa6730da2434b2b7f4cdc608acce0feefcf), it's fixed now

malfet commented:
> That's weird. I added ciflow/trunk too. Why CIs didn't catch this.

It's not a public build, but perhaps a good opportunity to rework PR a bit. Let's talk about it at 2PM today

malfet commented:
@pytorchbot

malfet commented:
Keeping PR closed for now in order to issue revert command

malfet commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -5,6 +5,15 @@
+#include <Foundation/Foundation.h>
+#include <Metal/Metal.h>
+typedef id<MTLCommandBuffer> MTLCommandBuffer_t;
+#else
+typedef void* MTLCommandBuffer_t;
+typedef void* dispatch_queue_t;
+#endif
```
Why not guard the entire header with `#ifdef __OBJC__` as it's clearly not usable unless OBJC is defined, is it?

razarmehr replied:
Only following APIs require ObjC, the rest are usable with C++:
torch::mps::get_command_buffer()
torch::mps::get_dispatch_queue()

malfet replied:
Sure, but my point is one can't really use ObjC Blocks or `MTLCommandBuffer_t` , so aliasing those to `void*` is kind of pointless.

razarmehr replied:
OK, I'll update the PR and will remove the typedefs.

razarmehr replied:
I updated the PR. I couldn't find any better way other than renaming the typedef of  `dispatch_queue_t` to `DispatchQueue_t`.

Two facts here:
1- We cannot guard the entire `mps.h` with `#ifdef __OBJC__`, since this header along with `mps.cpp` MUST build successfully on all platforms to be able to use APIs like `torch::mps::is_available()` on every platform for backend availability checking.
2- It's not possible to partially guard the code with `#ifdef __OBJC__` since mps.cpp is a CPP file and that condition will always be false. Also, we cannot rename it to mps.mm due to fact 1 above.

So, unless if you have a better suggestion that makes MPS's CPP APIs work everywhere, I think the solution I committed for `DispatchQueue_t` should be fine. Please let me know. Thanks.

malfet commented:
@pytorchbot revert -m "Break internal builds, but also guarding dispatch_t define behind __OBJC__ guard is not a good practices" -c ghfirst

razarmehr commented:
@pytorchbot rebase

razarmehr committed:
Resolve the naming conflict for dispatch_queue_t

razarmehr commented:
@pytorchbot rebase

malfet commented on torch/csrc/api/src/mps.cpp:
```diff
@@ -25,9 +21,22 @@ void manual_seed(uint64_t seed) {
+}
+
+MTLCommandBuffer_t get_command_buffer() {
+  return static_cast<MTLCommandBuffer_t>(
+      at::detail::getMPSHooks().getCommandBuffer());
```
Static cast is redundant here, isn't it?
```suggestion
  return at::detail::getMPSHooks().getCommandBuffer();
```

razarmehr replied:
The `getMPSHooks().getCommandBuffer()` returns a `void*` (since it comes from MPSHooksInterface.cpp which is CPP). So we need to cast it to the correct pointer type at this stage before returning it to user.

razarmehr replied:
Updated the PR and removed the explicit cast.

malfet commented on torch/csrc/api/src/mps.cpp:
```diff
@@ -25,9 +21,22 @@ void manual_seed(uint64_t seed) {
+}
+
+DispatchQueue_t get_dispatch_queue() {
+  return static_cast<DispatchQueue_t>(
+      at::detail::getMPSHooks().getDispatchQueue());
```
Same as above, this is just a pass-thru
```suggestion
  return at::detail::getMPSHooks().getDispatchQueue();
```

razarmehr replied:
Same as above, it's a void* that needs to be cast to `dispatch_queue_t` before returning

razarmehr replied:
Updated the PR.

malfet commented on torch/csrc/api/src/mps.cpp:
```diff
@@ -25,9 +21,22 @@ void manual_seed(uint64_t seed) {
   at::detail::getMPSHooks().deviceSynchronize();
 }
 
+void commit() {
+  at::detail::getMPSHooks().commitStream();
+}
```
Why not move this implementation to header and say that function is `inline`?

razarmehr replied:
I wanted the `MPSHooks` remain in the libtorch binary and not expose it.
But if you think that wouldn't make any problem with external users that only link their project against libtorch and use torch/mps.h header only (i.e., users who don't want to depend on any other torch headers like ATen/Context.h), then I can make those APIs static inline and expose MPSHooks in the mps.h header.

kulinseth replied:
@malfet , what do you think? are there any concerns.

TaiPhamD commented:
this a great addition. I have been reading the differences between implementing raw metal kernel vs using the MPSGraph api. In the current MPSGraph api has a concept of re-using Tensor cached graph to reduce allocation overhead. See here:
https://github.com/pytorch/pytorch/blob/036a8d6b4a77ff23185d00adb1128e6accadd022/aten/src/ATen/native/mps/operations/UnaryOps.mm#L73

Can we do something like this if we write raw metal kernel?

razarmehr committed:
Fix lint errors and remove static casts

razarmehr commented:
@pytorchbot rebase

kulinseth commented:
> this a great addition. I have been reading the differences between implementing raw metal kernel vs using the MPSGraph api. In the current MPSGraph api has a concept of re-using Tensor cached graph to reduce allocation overhead. See here:
> 
> https://github.com/pytorch/pytorch/blob/036a8d6b4a77ff23185d00adb1128e6accadd022/aten/src/ATen/native/mps/operations/UnaryOps.mm#L73
> 
> Can we do something like this if we write raw metal kernel?

Thanks @TaiPhamD. For raw Metal kernel you have the flexibility to cache the PSO as you see fit. We end up caching the Graph as there is CPU overhead in compiling it and we can do this as shapes are known. For Metal we already have caching built in-place for you in OS and you get that already.

malfet commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -5,6 +5,16 @@
+#ifdef __OBJC__
+#include <Foundation/Foundation.h>
+#include <Metal/Metal.h>
+typedef id<MTLCommandBuffer> MTLCommandBuffer_t;
+typedef dispatch_queue_t DispatchQueue_t;
+#else
+typedef void* MTLCommandBuffer_t;
+typedef void* DispatchQueue_t;
```
Please use `using` instead of `typedefs` in C++ headers.
```suggestion
using  MTLCommandBuffer_t = id<MTLCommandBuffer>;
using DispatchQueue_t =  dispatch_queue_t;
#else
using  MTLCommandBuffer_t = void*;
using  DispatchQueue_t = void*;
```

malfet committed:
Update torch/csrc/api/include/torch/mps.h

malfet committed:
Merge branch 'main' into MPS_CustomKernel

malfet commented on torch/csrc/api/include/torch/mps.h:
```diff
@@ -5,6 +5,16 @@
+#ifdef __OBJC__
+#include <Foundation/Foundation.h>
+#include <Metal/Metal.h>
+using  MTLCommandBuffer_t = id<MTLCommandBuffer>;
+using DispatchQueue_t =  dispatch_queue_t;
+#else
+using  MTLCommandBuffer_t = void*;
+using  DispatchQueue_t = void*;
```
```suggestion
using MTLCommandBuffer_t = id<MTLCommandBuffer>;
using DispatchQueue_t = dispatch_queue_t;
#else
using MTLCommandBuffer_t = void*;
using DispatchQueue_t = void*;
```

malfet committed:
Update torch/csrc/api/include/torch/mps.h

razarmehr commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -f "Internal builds are fine"
